### PR TITLE
use quickcache for domain get_by_name

### PR DESCRIPTION
**Pull request by snopoke**
Since this gets called many times per request it makes sense to cache it in local memory as well so that we don't have to hit the cache 20 times for the same thing in the same request.

@TylerSheffels

### CODE REVIEW

File: corehq/apps/domain/models.py [hunk 1]
```diff
@@ -517,6 +531,7 @@ def __unicode__(self):
         return self.name
 
     @classmethod
+    @quickcache(['name'], timeout=30*60, helper_class=_StrictQuickCache)
```
Comment (dimagimon) [inline corehq/apps/domain/models.py]: E226 missing whitespace around arithmetic operator

**Comment by snopoke** [comment]
@dannyroberts quickcache is great

**Comment by TylerSheffels** [comment]
:+1: after tests

File: corehq/apps/domain/models.py [hunk 2]
```diff
@@ -48,6 +49,19 @@
         lang_lookup[lang['two']] = lang['names'][0]
 
 
+class _StrictQuickCache(QuickCache):
```
Comment (dannyroberts) [inline corehq/apps/domain/models.py]: I like this, seems almost generalizable, but probably good idea to keep that out of scope for now.

One question—it seems like the original function would cache the response even in strict mode, and I think that's probably the right behavior. Does this also cache the response when strict=True?

File: corehq/apps/domain/models.py [hunk 3]
```diff
@@ -633,7 +639,7 @@ def case_sharing_included(self):
 
     def save(self, **params):
         super(Domain, self).save(**params)
-        cache.delete(_domain_cache_key(self.name))
+        self.get_by_name.clear(Domain, self.name)  # clear the domain cache
```
Comment (dannyroberts) [inline corehq/apps/domain/models.py]: Really nice! Glad to see a use of .clear in the wild lookin' so sharp!

**Comment by dannyroberts** [comment]
Great change, I love to see us consolidating around one way to do things. Left one comment, otherwise good to merge!

File: corehq/apps/domain/models.py (hunk 2 above)
Comment (snopoke) [inline corehq/apps/domain/models.py]: Great point about still caching, will update that.

Definitely think this is more generalizable but out of scope of this PR. Also going to change the name of `strict` to something better.

File: corehq/apps/domain/models.py [hunk 4]
```diff
@@ -48,6 +49,22 @@
...
+    """
+    def __call__(self, *args, **kwargs):
+        strict = kwargs.get('strict', False)
+        if not strict:
+            return super(_StrictQuickCache, self).__call__(*args, **kwargs)
+        else:
+            key = self.get_cache_key(*args, **kwargs)
+            content = self.fn(*args, **kwargs)
+            self.cache.set(key, content)
+            return content
```
Comment (snopoke) [inline corehq/apps/domain/models.py]: @dannyroberts updated to cache even in strict mode

Comment (dannyroberts) [inline corehq/apps/domain/models.py]: :+1:

**Comment by dannyroberts** [comment]
:x:  a little confused about this because the build failure looks pretty different, but could somehow be related? Same error on both commits

**Comment by czue** [comment]
late to the party, but :+1: to this change

**Comment by snopoke** [comment]
@dannyroberts finally found the issue here after lengthy debugging. quickcache differentiates between `str` and `unicode` vary on args and in some places `Domain.get_by_name` is being called with `str` and others with `unicode`.

Notably here its `str` which is coming straight out of the DB: https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/locations/models.py#L91-91

Not entirely sure what the best approach to fixing this is.

**Comment by dannyroberts** [comment]
@snopoke first of all, really impressed you found this as it's pretty unintuitive! There are a couple ways forward as I see it, but I'm also curious to know more

- How was the test relying on this caching nuance? Is it that it was expecting the domain object to be cached, and instead it returned a different python object (representing the same domain) because it wasn't cached?

Here are some options:
- explicit: have quickcache accept an arg definition like `"domain.__str__()"` (which would just require recognizing the `()` token and translating it to an arg-less fn call).
  - Plus: it's explicit—which means we can still differentiate between `str` and `unicode` when that distinction matters
  - Plus: it expands quickcache in a way that's probably useful in other scenarios too (calling an argless method)
  - Minus: it gets the ball rolling (uphill) for a sisyphean expansion of what quickcache accepts whose logical conclusion is reimplementing all of python to parse the `vary_on`
  - Minus: it's easy to write a cache that looks like it works but treats u'mydomain' and 'mydomain' as different inputs
- implicit: have quickcache treat `str` and `unicode` as the same type
  - Plus: this is almost always what people mean
  - Plus: it's easier to implement
  - Minus: you can no longer use it to cache a function that does care about the difference (e.g. `type`!)—but maybe who cares
- implicit, with a strict option: have treating `str` and `unicode` as equivalent be the default option, but have let the caller specify, e.g. `@quickcache(['value'], lenient_strings=False)`
  - Plus: all the pluses of the previous option, without the minus
  - Minus: slightly, slightly harder to implement
  - Minus: one more distinction that has to be tested and maintained

I think of these I'm leaning toward the third as a potential long term solution, and the second—plus clear comments on the behavior of str/unicode arguments—as the best immediate option. If we ever run into it, we can implement, and test, it then, with very little difficulty. I'm pretty sure no quickcache'd function relies on the distinguishing behavior currently.

**Comment by snopoke** [comment]
I agree with your options and also think that 2 is likely the right option
for now with 3 down the line if we find a need for it.

The way this was breaking was interesting because at the end of the test
you end up with two cache keys both referencing the same domain object. The
`tearDownClass` method calls `Domain.get_by_name('name').delete()` which
fetches the domain from the string cache key but clears the unicode key
(`self.domain`) leaving the string key which now points to a deleted
domain. When the next test runs this string key is still in the cache and
so calls to `get_by_name` with a string arg will return the deleted domain.

**Comment by dannyroberts** [comment]
Brilliant!

Looks like the recent changes is failing (differently):

```
======================================================================
ERROR: test_simple (corehq.apps.hqcase.tests.test_explode_cases.ExplodeCasesDbTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/dimagi/commcare-hq/corehq/apps/hqcase/tests/test_explode_cases.py", line 131, in test_simple
    explode_cases(self.user_id, self.domain.name, 10)
  File "/home/travis/build/dimagi/commcare-hq/corehq/apps/hqcase/tasks.py", line 16, in explode_cases
    keys = [[domain, owner_id, False] for owner_id in user.get_owner_ids()]
  File "/home/travis/build/dimagi/commcare-hq/corehq/apps/users/models.py", line 1839, in get_owner_ids
    if self.project.locations_enabled and self.location:
AttributeError: 'dict' object has no attribute 'locations_enabled'
```

and one more similar one. Looks like domains are getting loaded as `dict`s instead of `Domain`s.

**Comment by dannyroberts** [comment]
:x: I think there's the same issue happening again

**Comment by dannyroberts** [comment]
This is such an undeniably good change... sorry it's been so much more annoying to implement than it seemed!

**Comment by snopoke** [comment]
No problem, hardly your fault. I see the tests are failing :(

**Comment by snopoke** [comment]
@dannyroberts got to the bottom of this: https://github.com/dimagi/commcare-hq/commit/738b4057515c15df8dc482e143927b3d994b9357

**Comment by dannyroberts** [comment]
yay!

**Pull request merged**